### PR TITLE
!!! TASK: Introduce first level content repository projection

### PR DESCRIPTION
**Pull request by mhsdesign**
Replaces the quirky `ContentRepositoryReadModelAdapterInterface` with a new `ContentGraphReadModelInterface` that is a 1st level dependency to the `ContentRepository`.
This simplifies code flow and prevents CR instances without its main projection to be created.

## Note

This is a breaking change because it changes the configuration format for the `ContentRepositoryRegistry`  package from
```yaml
Neos:
  ContentRepositoryRegistry:
    presets:
      'default':
        # ...
        projections:
          'Neos.ContentRepository:ContentGraph':
            catchUpHooks: # ...
          'Neos.Neos:DocumentUriPathProjection': #...
```

to

```yaml
Neos:
  ContentRepositoryRegistry:
    presets:
      'default':
        # ...
        contentGraphProjection:
          catchUpHooks: #...
        projections:
          'Neos.Neos:DocumentUriPathProjection': #...
```

### CODE REVIEW

File: Neos.ContentRepository.Core/Classes/Projection/ProjectionsAndCatchUpHooks.php [hunk 1]
```diff
@@ -13,7 +13,8 @@
      * @param array<class-string<ProjectionInterface<ProjectionStateInterface>>, CatchUpHookFactories> $catchUpHookFactoriesByProjectionClassName
      */
     public function __construct(
-        public Projections $projections,
+        public ContentRepositoryReadModelProjection $readModelProjection,
+        public Projections $additionalProjections,
         private array $catchUpHookFactoriesByProjectionClassName,
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Projection/ProjectionsAndCatchUpHooks.php]: todo discuss naming of `        public ContentRepositoryReadModelProjection $readModelProjection,` and if we like this structure here in ProjectionAndCatchupHooks

also the construction of this in the factory is a litte hacky

File: Neos.ContentRepository.Core/Classes/ContentRepositoryReadModel.php [hunk 2]
```diff
@@ -26,118 +24,32 @@
...
+ * Create implementations of ContentGraphs bound to a specific Workspace and/or ContentStream
  *
- * The API way of accessing a ContentGraph is via ContentRepository::getContentGraph()
- *
- * @internal User land code should not use this directly.
- * @see ContentRepository::getContentGraph()
+ * @internal This is just an implementation detail to be implemented by the specific adapters
  */
-final class ContentRepositoryReadModel implements ProjectionStateInterface
+interface ContentRepositoryReadModel extends ProjectionStateInterface
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/ContentRepositoryReadModel.php]: Interfaces must have the `Interface` suffix

File: Neos.ContentRepository.Core/Classes/ContentRepository.php [hunk 3]
```diff
@@ -69,6 +70,23 @@ final class ContentRepository
...
+
+    /**
+     * @var array<string, Workspace> Runtime cache for {@see Workspace} instances, indexed by their name
+     */
+    private array $workspaceInstancesByName = [];
+
+    /**
+     * @var array<string, ContentStream> Runtime cache for {@see ContentStream} instances, indexed by their name
+     */
+    private array $contentStreamInstancesById = [];
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/ContentRepository.php]: Are those new runtime caches related? I would prefer to leave them for a separate PR

File: Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/CRTestSuiteTrait.php [hunk 4]
```diff
@@ -154,7 +156,20 @@ public function workspaceHasStatus(string $rawWorkspaceName, string $status): vo
...
+        $contentRepositoryReadModelAccess = new class implements ContentRepositoryServiceFactoryInterface {
+            public ContentRepositoryReadModel|null $instance;
+            public function build(ContentRepositoryServiceFactoryDependencies $serviceFactoryDependencies): ContentRepositoryServiceInterface
+            {
+                $this->instance = $serviceFactoryDependencies->projectionsAndCatchUpHooks->readModelProjection->getState();
+                return new class implements ContentRepositoryServiceInterface
+                {
+                };
+            }
+        };
```
Comment (bwaidelich) [inline Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/CRTestSuiteTrait.php:170]: We could declare some custom Test CR Service that just exposes all `ContentRepositoryServiceFactoryDependencies` to make these less of a hack – but that can be done in a follow-up of course

**Review by bwaidelich — COMMENTED** [review]
Nice one and it's a good idea to move the `countNodes()` hack to the higher level interface.
I already added a couple of comments even though this is just a draft yet

File: Neos.ContentRepository.Core/Classes/ContentRepository.php (hunk 3 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php]: removed

File: Neos.ContentRepository.Core/Classes/Projection/ContentRepositoryReadModelProjection.php [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace Neos\ContentRepository\Core\Projection;
+
+use Neos\ContentRepository\Core\ContentRepositoryReadModel;
+
+/**
+ * @extends ProjectionInterface<ContentRepositoryReadModel>
+ * @api
+ */
+interface ContentRepositoryReadModelProjection extends ProjectionInterface
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Projection/ContentRepositoryReadModelProjection.php]:
```suggestion
interface ContentRepositoryProjectionInterface extends ProjectionInterface
```

File: Neos.ContentRepository.Core/Classes/ContentRepositoryReadModel.php [hunk 6]
```diff
@@ -25,65 +24,26 @@
...
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/ContentRepositoryReadModel.php]:
```suggestion
interface ContentRepositoryReadModelInterface extends ProjectionStateInterface
```

File: Neos.ContentRepository.Core/Classes/CommandHandlingDependencies.php [hunk 7]
```diff
@@ -38,8 +38,10 @@ final class CommandHandlingDependencies
      */
     private array $overriddenContentGraphInstances = [];
 
-    public function __construct(private readonly ContentRepository $contentRepository)
-    {
+    public function __construct(
+        private readonly ContentRepository $contentRepository,
+        private readonly ContentRepositoryReadModel $adapter
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/CommandHandlingDependencies.php]: "adapter" is not very specific, what about:
```suggestion
        private readonly ContentRepositoryReadModel $contentRepositoryReadModel
```

File: Neos.ContentRepository.Core/Classes/ContentRepository.php [hunk 8]
```diff
@@ -69,8 +70,9 @@ final class ContentRepository
      */
     private array $projectionStateCache;
 
-    private CommandHandlingDependencies $commandHandlingDependencies;
+    private ContentRepositoryReadModel $adapter;
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/ContentRepository.php]: Same here:
```suggestion
    private ContentRepositoryReadModel $contentRepositoryReadModel;
```

File: Neos.ContentRepository.Core/Classes/Projection/ProjectionsAndCatchUpHooks.php [hunk 9]
```diff
@@ -13,7 +13,8 @@
      * @param array<class-string<ProjectionInterface<ProjectionStateInterface>>, CatchUpHookFactories> $catchUpHookFactoriesByProjectionClassName
      */
     public function __construct(
-        public Projections $projections,
+        public ContentRepositoryReadModelProjection $readModelProjection,
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Projection/ProjectionsAndCatchUpHooks.php]: I don't really like that distinction here.. In this case we should not make a difference between the individual projections IMO

**Review by bwaidelich — COMMENTED** [review]
@mhsdesign This is a great start already, but I'm not yet super happy with the split in the CR core.
I have some rough ideas I would like to discuss with you at some point

File: Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php [hunk 10]
```diff
@@ -242,29 +245,40 @@ private function buildPropertySerializer(ContentRepositoryId $contentRepositoryI
...
-                throw InvalidConfigurationException::fromMessage('Projection factory object name for projection "%s" (content repository "%s") is not an instance of %s but %s.', $projectionName, $contentRepositoryId->value, ProjectionFactoryInterface::class, get_debug_type($projectionFactory));
+            $this->registerProjection($projectionOptions, $projectionsFactory, ProjectionFactoryInterface::class, $projectionName, $contentRepositoryId);
+        }
+        return $projectionsFactory;
+    }
+
+    /**
+     * @param class-string<ProjectionFactoryInterface<ProjectionInterface<ProjectionStateInterface>>> $expectedProjectionFactory
+     */
+    private function registerProjection(mixed $projectionOptions, ProjectionsAndCatchUpHooksFactory $projectionsFactory, string $expectedProjectionFactory, string $projectionName, ContentRepositoryId $contentRepositoryId): void
```
Comment (mhsdesign) [inline Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php]: a little verbose ... sadly

File: Neos.ContentRepository.Core/Classes/Factory/ProjectionsAndCatchUpHooksFactory.php [hunk 11]
```diff
@@ -71,9 +73,17 @@ public function build(ProjectionFactoryDependencies $projectionFactoryDependenci
                 $options,
             );
             $catchUpHookFactoriesByProjectionClassName[$projection::class] = $catchUpHookFactories;
-            $projectionsArray[] = $projection;
+            if ($projection instanceof ContentGraphProjectionInterface) {
+                $contentGraphProjection = $projection;
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Factory/ProjectionsAndCatchUpHooksFactory.php]: little bit dirty but what solls :D

File: Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphReadModelAdapter.php [hunk 12]
```diff
@@ -32,10 +32,9 @@
 use Neos\EventStore\Model\Event\Version;
 
 /**
- * @internal only used inside the
- * @see ContentRepositoryReadModel
+ * @internal
  */
-final readonly class ContentRepositoryReadModelAdapter implements ContentRepositoryReadModelAdapterInterface
+final readonly class ContentGraphReadModelAdapter implements ContentGraphReadModelInterface
```
Comment (bwaidelich) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphReadModelAdapter.php:37]: Just wondering: Why did you rename this?
That read model knows all about workspaces etc. and those are not really part of the content graph, are they?

File: Neos.ContentRepository.Core/Classes/CommandHandlingDependencies.php [hunk 13]
```diff
@@ -82,8 +85,11 @@ public function getContentGraph(WorkspaceName $workspaceName): ContentGraphInter
         if (isset($this->overriddenContentGraphInstances[$workspaceName->value])) {
             return $this->overriddenContentGraphInstances[$workspaceName->value];
         }
-
-        return $this->contentRepository->getContentGraph($workspaceName);
+        $workspace = $this->contentGraphReadModel->findWorkspaceByName($workspaceName);
+        if ($workspace === null) {
+            throw WorkspaceDoesNotExist::butWasSupposedTo($workspaceName);
+        }
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/CommandHandlingDependencies.php:91]: With the [CR Privilges PR](https://github.com/neos/neos-development-collection/pull/5298) the evaluation of permissions is centralized to the [ContentRepository](https://github.com/neos/neos-development-collection/blob/bf57af6d351550860be19f28b8318889fc9d2836/Neos.ContentRepository.Core/Classes/ContentRepository.php#L250-L268)
I _think_ it's fine (or even desired) that they are _not_ evaluated here, but I'm not 100% sure

File: Neos.ContentRepository.Core/Classes/ContentRepository.php [hunk 14]
```diff
@@ -141,18 +143,24 @@ public function handle(CommandInterface $command): void
...
             return $projectionState;
         }
-        foreach ($this->projectionsAndCatchUpHooks->projections as $projection) {
-            $projectionState = $projection->getState();
-            if ($projectionState instanceof $projectionStateClassName) {
-                $this->projectionStateCache[$projectionStateClassName] = $projectionState;
-                return $projectionState;
-            }
+        if (in_array(ContentGraphReadModelInterface::class, class_implements($projectionStateClassName), true)) {
+            throw new \InvalidArgumentException(sprintf('Accessing the internal content repository projection state via %s(%s) is not allowed. Please use the API on the content repository instead.', __FUNCTION__, $projectionStateClassName), 1729338679);
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/ContentRepository.php:161]: nitpick: why not move this check to the top and keep the rest like before?

File: Neos.ContentRepository.Core/Classes/ContentRepository.php [hunk 15]
```diff
@@ -141,18 +143,24 @@ public function handle(CommandInterface $command): void
...
     {
+        if (!isset($this->projectionStateCache)) {
+            foreach ($this->projectionsAndCatchUpHooks->projections as $projection) {
+                if ($projection instanceof ContentGraphProjectionInterface) {
+                    continue;
+                }
+                $projectionState = $projection->getState();
+                $this->projectionStateCache[$projectionState::class] = $projectionState;
+            }
+        }
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/ContentRepository.php:154]: I don't completely get the implications of this, but why do we build up the runtime cache for all projections once any state is retrieved?

File: Neos.ContentRepository.Core/Classes/Projection/ProjectionStatuses.php [hunk 16]
```diff
@@ -18,9 +18,12 @@ private function __construct(
     ) {
     }
 
-    public static function create(): self
+    /**
+     * @param array<class-string<ProjectionInterface<ProjectionStateInterface>>, ProjectionStatus> $statuses
+     */
+    public static function create(array $statuses = []): self
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Projection/ProjectionStatuses.php]: nitpick: default values in method signatures can be error prone, I'd slighly prefer explicit constructors like
```php
public static function createEmpty(): self {...}
public static function fromArray(array $statuses): self {...}
```

**Review by bwaidelich — COMMENTED** [review]
First set of comments, looking great already.
What puzzles me is the distinction between content*Repository*Projection  (in configuration etc) but content*Graph*ReadModel

File: Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphReadModelAdapter.php (hunk 12 above)
Comment (kitsunet) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphReadModelAdapter.php:37]: They are?! We moved them to the contentgraph projection, IMHO because they are inherently part of it now?

Comment (kitsunet) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphReadModelAdapter.php:37]: I think you missed this yesterday then, sebastian and me suggested that this should all be ContentGraph instead of content repository.

File: Neos.ContentRepository.Core/Classes/ContentRepository.php (hunk 14 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php:161]: then it would be more expensive imo as we call `class_implements` always and here its just to improve the error message ;)

File: Neos.ContentRepository.Core/Classes/ContentRepository.php (hunk 15 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php:154]: because we did so previously almost as well just that we break on the first match, by building it once its more easy to grasp

File: Neos.ContentRepository.Core/Classes/ContentRepository.php [hunk 17]
```diff
@@ -141,18 +143,24 @@ public function handle(CommandInterface $command): void
      */
     public function projectionState(string $projectionStateClassName): ProjectionStateInterface
     {
+        if (!isset($this->projectionStateCache)) {
+            foreach ($this->projectionsAndCatchUpHooks->projections as $projection) {
+                if ($projection instanceof ContentGraphProjectionInterface) {
```
Comment (kitsunet) [inline Neos.ContentRepository.Core/Classes/ContentRepository.php:148]: So but that means you cannot get the CG projection here anymore? not sure that is good?!

File: Neos.ContentRepository.Core/Classes/ContentRepository.php (hunk 14 above)
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php:161]: as said above I would just allow it and thrash the check. it's a projectionstate after all, just one with _better_ API to reach it.

File: Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/CRTestSuiteTrait.php [hunk 18]
```diff
@@ -155,7 +157,20 @@ public function workspaceHasStatus(string $rawWorkspaceName, string $status): vo
      */
     public function iExpectTheGraphProjectionToConsistOfExactlyNodes(int $expectedNumberOfNodes): void
     {
-        $actualNumberOfNodes = $this->currentContentRepository->getContentGraph($this->currentWorkspaceName)->countNodes();
+        // HACK to access
```
Comment (kitsunet) [inline Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/CRTestSuiteTrait.php:160]: YIKES

File: Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php [hunk 19]
```diff
@@ -242,29 +245,40 @@ private function buildPropertySerializer(ContentRepositoryId $contentRepositoryI
...
+    {
+        (is_array($projectionOptions)) || throw InvalidConfigurationException::fromMessage('Projection "%s" (content repository "%s") must be configured as array got %s', $projectionName, $contentRepositoryId->value, get_debug_type($projectionOptions));
```
Comment (kitsunet) [inline Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php]: `(is_array($projectionOptions)) || throw InvalidConfigu...`

YIKES

also why is it mixed in the signature then

File: Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php [hunk 20]
```diff
@@ -242,29 +245,40 @@ private function buildPropertySerializer(ContentRepositoryId $contentRepositoryI
...
+        $projectionFactory = isset($projectionOptions['factoryObjectName']) ? $this->objectManager->get($projectionOptions['factoryObjectName']) : null;
+        if (!$projectionFactory instanceof $expectedProjectionFactory) {
```
Comment (kitsunet) [inline Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php]: Why do we need this here at all, aka why  do we need $expectedProjectionFactory?

File: Neos.ContentRepository.Core/Classes/ContentRepository.php (hunk 14 above)
Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php:161]: We will have to disallow this with the [CR Privileges](https://github.com/neos/neos-development-collection/pull/5298) because they enforce the constraints on the CR instance

File: Neos.ContentRepository.Core/Classes/ContentRepository.php (hunk 17 above)
Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php:148]: I think that's a _good_ thing (see comment above re PR privileges)

File: Neos.ContentRepository.Core/Classes/ContentRepository.php (hunk 14 above)
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php:161]: right, ok fair enough. Then this place is fine for me.

File: Neos.ContentRepository.Core/Classes/Factory/ProjectionsAndCatchUpHooksFactory.php (hunk 11 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Factory/ProjectionsAndCatchUpHooksFactory.php]: todo throw here if `$contentGraphProjection` already set!

File: Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php (hunk 19 above)
Comment (mhsdesign) [reply Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php]: later alligator

File: Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php (hunk 20 above)
Comment (mhsdesign) [reply Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php]: refactored

**Review by bwaidelich — APPROVED** [review]

**Review by bwaidelich — APPROVED** [review]
